toolchain/lower/mangler.cpp

```diff
@@ -0,0 +1,118 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/lower/mangler.h"
+
+#include "toolchain/base/kind_switch.h"
+#include "toolchain/sem_ir/entry_point.h"
+
+namespace Carbon::Lower {
+
+auto Mangler::MangleInverseQualifiedNameScope(llvm::raw_ostream& os,
+                                              SemIR::NameScopeId name_scope_id)
+    -> void {
+  // Maintain a stack of names for delayed rendering of interface impls.
+  struct NameEntry {
+    SemIR::NameScopeId name_scope_id;
+
+    // The prefix emitted before this name component. If '\0', no prefix will be
+    // emitted.
+    // - Namespace components are separated by '.'.
+    // - The two components of an interface are separated by ':'.
+    char prefix;
+  };
+  llvm::SmallVector<NameEntry> names_to_render;
+  names_to_render.push_back({.name_scope_id = name_scope_id, .prefix = '.'});
+  while (!names_to_render.empty()) {
+    auto [name_scope_id, prefix] = names_to_render.back();
+    names_to_render.pop_back();
+    if (prefix) {
+      os << prefix;
+    }
+    if (name_scope_id == SemIR::NameScopeId::Package) {
+      if (auto package_id = sem_ir().package_id(); package_id.is_valid()) {
+        os << sem_ir().identifiers().Get(package_id);
+      } else {
+        os << "Main";
+      }
+      continue;
+    }
+    const auto& name_scope = sem_ir().name_scopes().Get(name_scope_id);
+    CARBON_KIND_SWITCH(sem_ir().insts().Get(name_scope.inst_id)) {
+      case CARBON_KIND(SemIR::ImplDecl impl_decl): {
+        names_to_render.clear();
```

Comment:
Is this clear needed? Looks like you moved the push to the end, so maybe it's not needed anymore?

Note I have in mind something like:

```
class C {
  extend impl as I {
    fn F() {
      class C { extend impl as I { ... } }
    }
  }
}
```

Which seems like a terrible idea, but at the same time, valid code.

```suggestion
```

Reply:
Removed the `clear()` call - thanks for the catch.

Though the example code doesn't compile at the moment (maybe local classes aren't supported yet?
```
basic.carbon:7:7: ERROR: Expected expression.
      class C {
      ^~~~~
```

Though once supported, the mangling code will need to be enhanced to handle intermediate function names, not just at the top level.

Without considering local types, I was thinking about nested types and when dependent types get added to interfaces/impls, that might present similar problems depending on exactly how they work. Then we might have a case where an impl is a scope, though I guess it'd still be the top level scope... hmm, but might've bifurcated earlier on - at that point we might want some way to reference existing entities in the mangling, like Itanium...

Reply:
Yeah, not yet supported.

Reply:
I couldn't think of a case with existing code either, due to the restrictions on `impl as`